### PR TITLE
feat(compaction): cut file by table

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

close https://github.com/risingwavelabs/risingwave/issues/5299

This PR only implement the strategy that allow picker select sst files which have the same state-table-ids. I would create an independent compact-group to compact files splitting by state-table. The default state-group will keep all state-table in one file. We will move large state-table to another group so that we would not generate too much small file.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by Little-Wallace** [comment]
image
image

The left is current PR while the right is main branch. We can see that after compact by table we save some resource and could make trivial move more effective.

File: src/meta/src/hummock/compaction/picker/tier_compaction_picker.rs [hunk 1]
```diff
@@ -50,12 +50,134 @@ impl TierCompactionPicker {
...
+            let mut several_table = false;
+            for sst in &level.table_infos {
+                if sst.table_ids.len() > 1 {
+                    several_table = true;
+                    break;
+                }
+                if level_handler.is_pending_compact(&sst.id) {
+                    continue;
+                }
+                table_id = sst.table_ids[0];
```
Comment (Li0k) [inline src/meta/src/hummock/compaction/picker/tier_compaction_picker.rs:69]: Why do we choose the table_id of the last sst as the identifier for this task? The same concern whether starvation will occur

File: src/meta/src/hummock/compaction/picker/tier_compaction_picker.rs [hunk 2]
```diff
@@ -170,6 +270,12 @@ impl TierCompactionPicker {
                 continue;
             }
 
+            if select_tables.iter().map(|sst| sst.file_size).sum::<u64>()
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/meta/src/hummock/compaction/picker/tier_compaction_picker.rs:245]: Yes. I do not want to run too many trivial-move task.

File: src/meta/src/hummock/compaction/picker/base_level_compaction_picker.rs [hunk 3]
```diff
@@ -78,79 +73,117 @@ impl CompactionPicker for LevelCompactionPicker {
...
+            return Some(CompactionInput {
+                input_levels,
+                target_level: self.target_level,
+                target_sub_level_id: 0,
             });
-
-            let mut l0_total_file_size = l0.sub_levels[0].total_file_size;
-            for level in l0.sub_levels[1..].iter() {
+        }
+        for table_id in levels.member_table_ids.clone() {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/meta/src/hummock/compaction/picker/base_level_compaction_picker.rs]: I have add table id of all sst files to this set.

File: src/storage/hummock_sdk/src/key.rs [hunk 4]
```diff
@@ -732,6 +742,17 @@ mod tests {
         let key = FullKey::for_test(TableId::new(1), &table_key[..], 1);
         let buf = key.encode();
         assert_eq!(FullKey::decode(&buf), key);
+        let mut table_key = vec![1];
+        let a = FullKey::for_test(TableId::new(1), table_key.clone(), 1);
+        println!("a: {}", a.user_key.get_vnode_id());
```
Comment (Li0k) [inline src/storage/hummock_sdk/src/key.rs:747]: remove this?

**Review by Li0k — APPROVED** [review]
Rest LGTM, thanks for the PR

**Pull request merged**